src/java.net.http/share/classes/jdk/internal/net/http/LimitingSubscriber.java

```diff
@@ -0,0 +1,150 @@
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.internal.net.http;
+
+import jdk.internal.net.http.ResponseSubscribers.TrustedSubscriber;
+import jdk.internal.net.http.common.Utils;
+
+import java.io.IOException;
+import java.net.http.HttpResponse.BodySubscriber;
+import java.nio.Buffer;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Flow.Subscription;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A subscriber limiting the maximum number of bytes that are allowed to be consumed by a downstream subscriber.
+ *
+ * @param <T> the response type
+ */
+public final class LimitingSubscriber<T> implements TrustedSubscriber<T> {
+
+    private final BodySubscriber<T> downstreamSubscriber;
+
+    private final long capacity;
+
+    private final AtomicReference<State> stateRef = new AtomicReference<>();
+
+    private long length;
+
+    private interface State {
+
+        enum Terminated implements State { INSTANCE }
+
+        record Subscribed(Subscription subscription) implements State {}
+
+    }
+
+    /**
+     * @param downstreamSubscriber the downstream subscriber to pass received data to
+     * @param capacity the maximum number of bytes that are allowed
+     * @throws IllegalArgumentException if {@code capacity < 0}
+     */
+    public LimitingSubscriber(BodySubscriber<T> downstreamSubscriber, long capacity) {
+        if (capacity < 0) {
+            throw new IllegalArgumentException("was expecting \"capacity >= 0\", found: " + capacity);
+        }
+        this.downstreamSubscriber = requireNonNull(downstreamSubscriber, "downstreamSubscriber");
+        this.capacity = capacity;
+    }
+
+    @Override
+    public void onSubscribe(Subscription subscription) {
+        requireNonNull(subscription, "subscription");
+        boolean subscribed = stateRef.compareAndSet(null, new State.Subscribed(subscription));
+        if (!subscribed) {
+            subscription.cancel();
+        } else {
+            downstreamSubscriber.onSubscribe(subscription);
+        }
+    }
+
+    @Override
+    public void onNext(List<ByteBuffer> buffers) {
+
+        // Check arguments
+        requireNonNull(buffers, "buffers");
+        assert Utils.hasRemaining(buffers);
+
+        // Short-circuit if not subscribed
+        if (!(stateRef.get() instanceof State.Subscribed subscribed)) {
+            return;
+        }
+
+        // See if we may consume the input
+        if (allocateLength(buffers)) {
+            downstreamSubscriber.onNext(buffers);
+        }
+
+        // Otherwise, trigger failure
+        else if (stateRef.compareAndSet(subscribed, State.Terminated.INSTANCE)) {
```

Comment:
I would prefer to keep the regular style:

```
if (condition) {
    // do something
} else {
    // do something else
}
```

Reply:
Fixed in bb577e9ba8a6ce8ee24e8371c4e34aeaceeb6808.